File: server/src/routes/graph.js
import express from "express";
import { HumanMessage, AIMessage } from "@langchain/core/messages";
import { Command } from "@langchain/langgraph";
import { buildCustomerGraph } from "../graphs/customer-graph.js";
import { messageDB } from "../db/postgres.js";
import { randomUUID } from "crypto";

const router = express.Router();

// 懒加载图实例（checkpointer 需要异步初始化）
let graphInstance = null;

const getGraph = async () => {
  if (!graphInstance) {
    graphInstance = await buildCustomerGraph();
  }
  return graphInstance;
};

// SSE 响应头设置
const setSSEHeaders = (res) => {
  res.setHeader("Content-Type", "text/event-stream; charset=utf-8");
  res.setHeader("Cache-Control", "no-cache");
  res.setHeader("Connection", "keep-alive");
  res.setHeader("X-Accel-Buffering", "no");
};

// ─── 图流式对话 ──────────────────────────────────────────────────
router.post("/stream", async (req, res) => {
  const { message, history = [], sessionId } = req.body;

  if (!message) return res.status(400).json({ error: "message 不能为空" });

  setSSEHeaders(res);

  const send = (type, data) =>
    res.write(`data: ${JSON.stringify({ type, ...data })}\n\n`);

  try {
    const g = await getGraph();

    // 构建 thread_id（= session_id），用于 checkpoint 标识
    const threadId = sessionId || randomUUID();

    const historyMessages = history.map((m) =>
      m.role === "user"
        ? new HumanMessage(m.content)
        : new AIMessage(m.content),
    );

    // 用户消息入库
    if (sessionId) {
      try {
        await messageDB.add(sessionId, "user", message);
      } catch (err) {
        console.error("[Graph Message Persist Error]", err.message);
      }
    }

    const stream = await g.stream(
      {
        userInput: message,
        messages: [...historyMessages, new HumanMessage(message)],
      },
      {
        streamMode: "updates",
        configurable: { thread_id: threadId },
      },
    );

    let assistantResponse = "";

    for await (const update of stream) {
      const [nodeName, nodeState] = Object.entries(update)[0];

      send("node", {
        node: nodeName,
        intent: nodeState.intent || null,
        intents: nodeState.intents || null, // 多意图数组
      });

      // 安全护栏拦截事件
      if (nodeState.guardrailBlocked) {
        send("block", { reason: "内容安全检查未通过" });
      }

      if (nodeState.orderResult?.steps?.length) {
        send("steps", { steps: nodeState.orderResult.steps });
      }

      // 转发 RAG 参考来源
      if (nodeState.ragSources?.length) {
        send("sources", { sources: nodeState.ragSources });
      }

      if (nodeState.finalAnswer) {
        assistantResponse = nodeState.finalAnswer;
        send("answer", { content: nodeState.finalAnswer });
      } else if (nodeState.orderResult?.answer) {
        assistantResponse = nodeState.orderResult.answer;
        send("answer", { content: nodeState.orderResult.answer });
      } else if (nodeState.ragResult) {
        assistantResponse = nodeState.ragResult;
        send("answer", { content: nodeState.ragResult });
      }
    }

    // 检查图是否被中断（HITL 等待审批）
    const state = await g.getState({ configurable: { thread_id: threadId } });
    const tasks = state?.tasks || [];
    const hasInterrupt = tasks.some((t) => t.interrupts?.length > 0);

    if (hasInterrupt) {
      // 图被中断 — 发送 interrupt 事件，不发送 done
      const interruptTask = tasks.find((t) => t.interrupts?.length > 0);
      const interruptInfo = interruptTask.interrupts[0].value;

      send("interrupt", {
        threadId,
        ...interruptInfo,
      });
      res.end();
      return; // 不入库，等待 resume
    }

    // AI 回复入库
    if (sessionId && assistantResponse) {
      try {
        await messageDB.add(sessionId, "assistant", assistantResponse, {
          mode: "Graph",
          threadId,
        });
      } catch (err) {
        console.error("[Graph Message Persist Error]", err.message);
      }
    }

    send("done", {});
    res.end();
  } catch (err) {
    console.error("[Graph Error]", err.message);
    send("error", { content: "处理请求时出错，请重试" });
    res.end();
  }
});

// ─── 恢复被中断的图执行 ─────────────────────────────────────
router.post("/resume", async (req, res) => {
  const { threadId, decision, sessionId } = req.body;

  if (!threadId) return res.status(400).json({ error: "threadId 不能为空" });
  if (!decision) return res.status(400).json({ error: "decision 不能为空" });

  setSSEHeaders(res);

  const send = (type, data) =>
    res.write(`data: ${JSON.stringify({ type, ...data })}\n\n`);

  try {
    const g = await getGraph();

    // 使用 Command.resume 恢复图执行
    const stream = await g.stream(new Command({ resume: decision }), {
      streamMode: "updates",
      configurable: { thread_id: threadId },
    });

    let assistantResponse = "";

    for await (const update of stream) {
      const [nodeName, nodeState] = Object.entries(update)[0];

      send("node", { node: nodeName, intent: nodeState.intent || null });

      if (nodeState.orderResult?.steps?.length) {
        send("steps", { steps: nodeState.orderResult.steps });
      }

      // 转发 RAG 参考来源
      if (nodeState.ragSources?.length) {
        send("sources", { sources: nodeState.ragSources });
      }

      if (nodeState.finalAnswer) {
        assistantResponse = nodeState.finalAnswer;
        send("answer", { content: nodeState.finalAnswer });
      } else if (nodeState.orderResult?.answer) {
        assistantResponse = nodeState.orderResult.answer;
        send("answer", { content: nodeState.orderResult.answer });
      } else if (nodeState.ragResult) {
        assistantResponse = nodeState.ragResult;
        send("answer", { content: nodeState.ragResult });
      }
    }

    // AI 回复入库
    if (sessionId && assistantResponse) {
      try {
        await messageDB.add(sessionId, "assistant", assistantResponse, {
          mode: "Graph",
          threadId,
          hitl: decision,
        });
      } catch (err) {
        console.error("[Graph Resume Persist Error]", err.message);
      }
    }

    send("done", {});
    res.end();
  } catch (err) {
    console.error("[Graph Resume Error]", err.message);
    send("error", { content: "恢复执行时出错，请重试" });
    res.end();
  }
});

export default router;

File: backend_py/routes/graph.py
"""
LangGraph 工作流路由
对照 JS: router.post("/stream", ...) + g.stream({...}, { streamMode: "updates" })
"""
import json
from fastapi import APIRouter
from fastapi.responses import StreamingResponse
from pydantic import BaseModel
from langchain_core.messages import HumanMessage, AIMessage
from graphs.customer_graph import build_customer_graph

router = APIRouter(prefix="/api/graph", tags=["graph"])

# 单例：避免每次请求都重新编译图
_graph = None


def get_graph():
    """对照 JS: let graph = null; const getGraph = () => { if (!graph) graph = buildCustomerGraph(); return graph; }"""
    global _graph
    if _graph is None:
        _graph = build_customer_graph()
    return _graph


class GraphRequest(BaseModel):
    message: str
    history: list = []


@router.post("/stream")
async def graph_stream(req: GraphRequest):
    if not req.message:
        return {"error": "message 不能为空"}

    async def generate():
        try:
            g = get_graph()

            # 把历史消息转成 LangChain Message 格式
            history_messages = []
            for m in req.history:
                if m.get("role") == "user":
                    history_messages.append(HumanMessage(content=m["content"]))
                elif m.get("role") == "assistant":
                    history_messages.append(AIMessage(content=m["content"]))

            # 流式执行图，每个节点执行完毕后触发一次事件
            # 对照 JS: const stream = await g.stream({...}, { streamMode: "updates" })
            # Python:  async for update in g.astream({...}, stream_mode="updates")
            async for update in g.astream(
                {
                    "userInput": req.message,
                    "messages": [*history_messages, HumanMessage(content=req.message)],
                },
                stream_mode="updates",
            ):
                # update 是一个字典 {nodeName: nodeState}
                for node_name, node_state in update.items():
                    # 推送节点执行事件
                    yield f"data: {json.dumps({'type': 'node', 'node': node_name, 'intent': node_state.get('intent')}, ensure_ascii=False)}\n\n"

                    # 推送订单步骤
                    order_result = node_state.get("orderResult")
                    if order_result and isinstance(order_result, dict) and order_result.get("steps"):
                        yield f"data: {json.dumps({'type': 'steps', 'steps': order_result['steps']}, ensure_ascii=False)}\n\n"

                    # 推送最终答案
                    final_answer = node_state.get("finalAnswer")
                    if final_answer:
                        yield f"data: {json.dumps({'type': 'answer', 'content': final_answer}, ensure_ascii=False)}\n\n"

            yield f"data: {json.dumps({'type': 'done'})}\n\n"
        except Exception as e:
            print(f"[Graph Error] {e}")
            yield f"data: {json.dumps({'type': 'error', 'content': '处理请求时出错，请重试'}, ensure_ascii=False)}\n\n"

    return StreamingResponse(
        generate(),
        media_type="text/event-stream",
        headers={
            "Cache-Control": "no-cache",
            "Connection": "keep-alive",
            "X-Accel-Buffering": "no",
        },
    )